Introduce server plugin helpers

tagomoris opened the pull request:
Server plugin helpers are to make plugin authors to write network servers very easily.
These make authors free from writing code to create sockets, handle event loops or threads.

The key features are:
* callback-based network daemon execution (plugin helper style)
* test driver support to wait startup/shutdown servers before actual test code running
* creating sockets via socket manager in default for coming multi-process configuration support

The full features of socket/server plugin helpers and TBDs are:
* TCP, UDP and TLS support (TLS will be implemented in future change)
* TCP/TLS keepalive (in future)
* IPv6 support

I committed all changes for this pull-request
* server plugin helper: TCP/UDP implementation complete
* tcp/udp input plugin: migrated to v0.14 APIs with server plugin helper
* forward input plugin: updated with server plugin helper
* tests of server plugin helper

To reduce the size of diff, keepalive support and socket plugin helper support will be implemented in following pull requests.
* socket plugin helper and its tests
* update of output plugins using sockets (including out_forward)
* http input plugin update (in future with keepalive support)

tagomoris commented:
@repeatedly Could you take a look for API design and basic features? If you are ok, I'll write tests and move forward to implement socket plugin helper.

repeatedly commented on lib/fluent/plugin/in_forward.rb:
```diff
@@ -315,29 +310,31 @@ def on_message(msg, chunk_size, peeraddr)
       # return option for response
       option
+    ensure
+      p(here: "ensure of on_message", error: $!) if $!
```
Why use `p`?

tagomoris replied:
It's just about debugging. I missed to remote it.

repeatedly commented on lib/fluent/plugin_helper/server.rb:
```diff
@@ -0,0 +1,375 @@
+        end
+
+        def initialize(sock, resolve_name, max_bytes, flags, log, under_plugin_development, &callback)
+          raise ArgumentError, "socket is a UDPSocket" unless sock.is_a?(UDPSocket)
```
`socket must be a UDPSocket: socket = #{sock}` or `socket is not a UDPSocket: socket = #{sock}`  is better.

repeatedly commented:
Basic implementation looks good.

Off topic: How about moving `under_plugin_development` into Engine like `Engine.under_plugin_development?`
Passing `under_plugin_development` to each method seems bit messy...

repeatedly commented on lib/fluent/plugin_helper/server.rb:
```diff
@@ -0,0 +1,375 @@
+        def on_read_without_connection(data)
+          @data_callback.call(data)
+        rescue => e
+          p(here: "error without connection", error: e)
```
`p` debug.

repeatedly commented on lib/fluent/plugin_helper/server.rb:
```diff
@@ -0,0 +1,375 @@
+        def on_read_with_connection(data)
+          @data_callback.call(data, self)
+        rescue => e
+          p(here: "error with connection", error: e)
```
ditto

tagomoris commented:
@repeatedly No way. It's very bad practice to use Engine for such purpose. Engine is a singleton object, and using Engine for such purpose makes many messy code to overriding `Engine.under_plugin_development?` and to revert overriding in tests.

tagomoris commented:
I pushed updated commits to follow review comments, and add some changes:
* add `Fluent::PluginHelper::Server::EventHandler` module and added TCPServer/UDPServer under it
* add `Fluent::PluginHelper::Server::CallbackSocket` and its subclasses of it for tcp/udp

I changed to pass an instance object of CallbackSocket to callbacks, to avoid to pass instances of cool.io classes. If plugin helper passes instances of cool.io classes, users may use methods of these classes (even if we want to keep these methods as semi-private).
For instances of original classes, we can control the variations of these methods (for compatibility).

tagomoris commented:
There's no change about server plugin helper API for plugins in updated commits.

tagomoris commented:
I pushed all commits to add/fix server plugin helper, and tests which pass on my laptop.
I'll check CI failure and fix it, but anyway, @repeatedly could you review this change?

repeatedly commented on lib/fluent/plugin/in_tcp.rb:
```diff
@@ -14,28 +14,66 @@
+              tag = extract_tag_from_record(record)
+              tag ||= @tag
+              time ||= extract_time_from_record(record) || Fluent::EventTime.now
+              record[@source_host_key] = conn.remote_host if @source_host_key
```
use `@source_hostname_key` instead.

repeatedly commented on lib/fluent/plugin/in_udp.rb:
```diff
@@ -14,24 +14,58 @@
+            tag = extract_tag_from_record(record)
+            tag ||= @tag
+            time ||= extract_time_from_record(record) || Fluent::EventTime.now
+            record[@source_host_key] = sock.remote_host if @source_host_key
```
ditto

repeatedly commented on lib/fluent/plugin_helper/event_loop.rb:
```diff
@@ -30,12 +30,16 @@ module EventLoop
       def event_loop_attach(watcher)
         @_event_loop_mutex.synchronize do
           @_event_loop.attach(watcher)
+          @_event_loop_attached_watchers << watcher
```
This means `Loop#watchers` sometimes returns incorrect watchers?

tagomoris replied:
`Loop#watchers` contains watchers which are NOT registered by event loop helpers, like TCP connection I/O watchers registered by TCP server watcher.
These watchers should be detached/closed by these watchers itself. Detaching such watchers by event loop helpers forcedly makes troubles.

repeatedly commented on lib/fluent/plugin_helper/server.rb:
```diff
@@ -0,0 +1,494 @@
+          when :data
+            @sock.data(&callback)
+          when :write_complete
+            cb = ->(){
```
Use one line for `:close` code.

repeatedly commented on lib/fluent/plugin_helper/server.rb:
```diff
@@ -0,0 +1,494 @@
+          SOCK_OPT_FORMAT = 'I!I!' # { int l_onoff; int l_linger; }
+
+          def initialize(sock, close_callback, resolve_name, linger_timeout, log, under_plugin_development, connect_callback)
+            raise ArgumentError, "socket must be a TCPSocket" unless sock.is_a?(TCPSocket)
```
add `: sock = #{sock}` to show actual object.

repeatedly commented:
New code/classes looks good.

nurse commented:
treasure-data/serverengine#82 fixes the test failure.

tagomoris commented:
@nurse @unak Great work, thank you!

tagomoris commented on lib/fluent/plugin/in_tcp.rb:
```diff
@@ -14,28 +14,66 @@
 
-    def listen(callback)
-      log.info "listening tcp socket on #{@bind}:#{@port}"
+    def configure(conf)
```
Right.

tagomoris commented:
I pushed some commits for review comments, and will merge after CI green.
